ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java

```diff
@@ -206,16 +209,16 @@ private InstallSnapshotReplyProto notifyStateMachineToInstallSnapshot(
       state.setLeader(leaderId, "installSnapshot");
       server.updateLastRpcTime(FollowerState.UpdateType.INSTALL_SNAPSHOT_NOTIFICATION);
 
-      if (inProgressInstallSnapshotIndex.compareAndSet(0, firstAvailableLogIndex)) {
+      if (inProgressInstallSnapshotIndex.compareAndSet(INVALID_LOG_INDEX, firstAvailableLogIndex)) {
         LOG.info("{}: Received notification to install snapshot at index {}", getMemberId(), firstAvailableLogIndex);
         // Check if snapshot index is already at par or ahead of the first
         // available log index of the Leader.
         final long snapshotIndex = state.getLog().getSnapshotIndex();
-        if (snapshotIndex + 1 >= firstAvailableLogIndex && firstAvailableLogIndex > 0) {
+        if (snapshotIndex >= firstAvailableLogIndex && firstAvailableLogIndex > INVALID_LOG_INDEX) {
```

Comment:
snapshotIndex + 1 seems correct since snapshotIndex is the last index in a snapshot.  Then, leader can send log starting at firstAvailableLogIndex to this server.

Reply:
Do I understand correctly that `snapshotIndex` is the last index in a snapshot that follower already installed, and `firstAvailableLogIndex` is the first index available from leader that would be part of a new snapshot if follower requested?

If so, assume `snapshotIndex = 9` and `firstAvailableLogIndex = 10`.  In this case 9+1 >= 10, hence follower responds with `ALREADY_INSTALLED`.  Does that seem right?

Reply:
> Do I understand correctly that `snapshotIndex` is the last index in a snapshot that follower already installed, and `firstAvailableLogIndex` is the first index available from leader that would be part of a new snapshot if follower requested?
> 
> If so, assume `snapshotIndex = 9` and `firstAvailableLogIndex = 10`. In this case 9+1 >= 10, hence follower responds with `ALREADY_INSTALLED`. Does that seem right?

I think so. Hope @szetszwo  help confirm it

Reply:
Yes, ALREADY_INSTALLED is correct.

Leader tells firstAvailableLogIndex (=10) to a follower.   The follower checks if it needs to install a snapshot -- if its snapshotIndex < 9, say snapshotIndex == 8, there is a gap between snapshotIndex and firstAvailableLogIndex so that it needs to install a snapshot with index >= 9.  The leader is not able to send the index 9 log entry to the follower because firstAvailableLogIndex is 10.

If the follower's snapshotIndex == 9, there is no gap so that it can tell the leader to send log entires starting at index 10 without installing a snapshot.

Reply:
Thanks for the explanation.